File: java/src/net/mackenzie_serres/router/Optimal.java
package net.mackenzie_serres.router;

public class Optimal {
	static RoadSystem.Path calculate(RoadSystem system) {
		RoadSystem.Path[] bestPaths = new RoadSystem.Path[2];
		bestPaths[0] = new RoadSystem.Path();
		bestPaths[1] = new RoadSystem.Path();

		for (RoadSystem.Section section : system.getSections()) {
			roadStep(bestPaths, section);
		}

		if (bestPaths[0].totalTime() < bestPaths[1].totalTime())
			return bestPaths[0];
		else
			return bestPaths[1];
	}

	static void roadStep(RoadSystem.Path[] paths, RoadSystem.Section section) {
		int timeA = paths[0].totalTime();
		int timeB = paths[1].totalTime();
		int forwardTimeToA = timeA + section.a;
		int crossTimeToA = timeB + section.b + section.c;
		int forwardTimeToB = timeB + section.b;
		int crossTimeToB = timeA + section.a + section.c;

		RoadSystem.Path oldA = paths[0].clone();

		if (forwardTimeToA <= crossTimeToA)
			paths[0].add(new RoadSystem.Point(RoadSystem.Label.A, section.a));
		else {
			paths[0] = paths[1].clone();
			paths[0].add(new RoadSystem.Point(RoadSystem.Label.B, section.b));
			paths[0].add(new RoadSystem.Point(RoadSystem.Label.C, section.c));
		}

		if (forwardTimeToB <= crossTimeToB)
			paths[1].add(new RoadSystem.Point(RoadSystem.Label.B, section.b));
		else {
			paths[1] = oldA.clone();
			paths[1].add(new RoadSystem.Point(RoadSystem.Label.A, section.a));
			paths[1].add(new RoadSystem.Point(RoadSystem.Label.C, section.c));
		}
	}
}

File: go/src/optimal/optimal.go
package optimal

import "roads"

func Calculate(rs * roads.System) (path * roads.Path) {
	bestPaths := make([]roads.Path, 2)

	for _, section := range rs.GetSections() {
		roadStep(bestPaths, section)
	}

	if bestPaths[0].TotalTime() < bestPaths[1].TotalTime() {
		return &bestPaths[0]
	} else {
		return &bestPaths[1]
	}
}

func roadStep(paths []roads.Path, section roads.Section) {
	var timeA = paths[0].TotalTime()
	var timeB = paths[1].TotalTime()
	var forwardTimeToA = timeA + section.A
	var crossTimeToA = timeB + section.B + section.C
	var forwardTimeToB = timeB + section.B
	var crossTimeToB = timeA + section.A + section.C

	var oldA = paths[0]

	if (forwardTimeToA <= crossTimeToA) {
		paths[0].AddPoint(roads.NewPoint("A", section.A))
	} else {
		paths[0] = paths[1]
		paths[0].AddPoint(roads.NewPoint("B", section.B))
		paths[0].AddPoint(roads.NewPoint("C", section.C))
	}

	if (forwardTimeToB <= crossTimeToB) {
		paths[1].AddPoint(roads.NewPoint("B", section.B))
	} else {
		paths[1] = oldA
		paths[1].AddPoint(roads.NewPoint("A", section.A))
		paths[1].AddPoint(roads.NewPoint("C", section.C))
	}
}
